trino/auth.py

```diff
@@ -168,14 +176,24 @@ class _OAuth2TokenBearer(AuthBase):
     MAX_OAUTH_ATTEMPTS = 5
     _BEARER_PREFIX = re.compile(r"bearer", flags=re.IGNORECASE)
 
-    def __init__(self, http_session, redirect_auth_url_handler=handle_redirect_auth_url):
+    def __init__(self, redirect_auth_url_handler=handle_redirect_auth_url):
         self._redirect_auth_url = redirect_auth_url_handler
-        self._thread_local = threading.local()
-        http_session.hooks['response'].append(self._authenticate)
+        try:
+            self._keyring = importlib.import_module("keyring")
+        except ImportError:
+            self._keyring = None
+            print("Warning: keyring module not found. OAuth2 token will not be stored in keyring.")
+        self._token = None
+        self._token_lock = threading.Lock()
```

Comment:
Why there are 2 locks needed?

Reply:
I'm not sure I understand. So there are two critical sections:
* get/set token from the cache,
* obtain and replace the token.
For the latter `_inside_oauth_attempt_lock` does guarantee that only one thread tries to get a new token at a given time but I believe other threads will still try to get a new token after obtaining this lock even though the token was just refreshed and is still valid:
```python
with self._inside_oauth_attempt_lock:
    if not self._inside_oauth_attempt:
        self._inside_oauth_attempt = True
        self._attempt_oauth(response, **kwargs)
        self._inside_oauth_attempt = False
```

Reply:
You're indeed right, it's not correct. I will make the change and add a test for this behaviour.